extensions/amp-ad-exit/0.1/amp-ad-exit.js

```diff
@@ -190,6 +205,74 @@ export class AmpAdExit extends AMP.BaseElement {
       this.user().error(TAG, 'Invalid JSON config', e);
       throw e;
     }
+
+    this.ampAdResourceId_ = user().assert(
+        this.getAmpAdResourceId_(),
+        `${TAG}: No friendly parent amp-ad element was found for amp-ad-exit.`);
+
+    this.unlisten_ = listen(this.getAmpDoc().win, 'message', event => {
+      const responseMessage = deserializeMessage(getData(event));
+
+      this.assertValidResponseMessage_(responseMessage, event.origin);
+
+      this.vendorResponses_[responseMessage['vendor']] =
+          responseMessage['message'];
+    });
+  }
+
+  /**
+   * Gets the resource ID of the amp-ad element containing this amp-ad-exit tag.
+   * This is a pass-through for the version in service.js, solely because
+   * the one in service.js isn't stubbable for testing, since only object
+   * methods are stubbable.
+   * @returns {?string}
+   */
+  getAmpAdResourceId_() {
+    return getAmpAdResourceId(this.element, this.win.top);
+  }
+
+  /** @override */
+  detachedCallback() {
+    if (this.unlisten_) {
+      this.unlisten_();
+    }
+  }
+
+  /**
+   *
+   * @param {?JsonObject|undefined} responseMessage The response object to validate.
+   * @param {string} expectedVendor The 3p analytics vendor, which
+   *     responseMesssage['vendor'] should match.
+   * @private
+   */
+  assertValidResponseMessage_(responseMessage, expectedVendor) {
+    user().assert(responseMessage && responseMessage['message'],
+        'Received empty response from 3p analytics frame');
+    user().assert(
+        responseMessage['type'] == MessageType.IFRAME_TRANSPORT_RESPONSE,
+        'Received response message of invalid type from 3p analytics frame');
+    user().assert(
+        responseMessage['creativeId'] == this.ampAdResourceId_,
+        'Received malformed message from 3p analytics frame: ' +
+        'creativeId missing');
+    user().assert(responseMessage['vendor'],
+        'Received malformed message from 3p analytics frame: ' +
+        'vendor missing');
+    this.assertOriginMatchesVendor_(expectedVendor, responseMessage['vendor']);
+  }
+
+  /**
+   * Ensures that a given origin matches that of an existing vendor's
+   * transport/iframe URL
+   * @param {string} origin The origin to verify
+   * @param {string} vendor The vendor whose origin to check against
+   * @private
+   */
+  assertOriginMatchesVendor_(origin, vendor) {
+    assertVendor(vendor);
```

Comment:
consider having this return the value of ANALYTICS_CONFIG[vendor] so that you don't look it up twice (or need to import ANALYTICS_CONFIG here)

Reply:
Nice idea! Done.